### PR TITLE
[inductor] add evict-first

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119622

In https://github.com/pytorch/pytorch/issues/119523 we discovered that in persistent reduction, adding evict_first hint to loads that are not used later can help perf quite a bit for small kernels. E.g. for the example i added in test_evict_first_for_persistent_reduction , the perf improves by 16% for that kernel. I'm not sure how this affect large kernels. But I'll run a perf test.

Our code seems to on purpose skip adding these hints for pointwise / persistent reduction. Please comment if there is reason why we do that. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by shunting314** [comment]
Perf test running here: https://github.com/pytorch/pytorch/actions/runs/7851926430

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1877,7 +1875,7 @@ def load(self, name: str, index: sympy.Expr):
             ep = ", eviction_policy='evict_last'"
         elif not is_coalesced:
             ep = ", eviction_policy='evict_last'"
-        elif self.inside_reduction and self.range_trees[-1].is_loop:
+        elif self.inside_reduction:
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:1878]: This should probably apply to pointwise kernels as well?

**Review by jansel — APPROVED** [review]
Please confirm the results of the perf test in a comment before landing.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:1878]: Oh right, I was planning to apply this to both pointwise and persistent reduction. Thanks for catching this!

**Comment by Chillee** [comment]
@shunting314 I added the original evict_first heuristic, and I think I just did it to minimize the amount of cases it applies to. I had a particular use case for that (matrix-vector multiplications), and I didn't want it to make things worse for other cases.

https://github.com/pytorch/pytorch/pull/108841/files

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1877,7 +1875,7 @@ def load(self, name: str, index: sympy.Expr):
             ep = ", eviction_policy='evict_last'"
         elif not is_coalesced:
             ep = ", eviction_policy='evict_last'"
-        elif self.inside_reduction and self.range_trees[-1].is_loop:
+        else:
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1878]: Rather than doing this, we probably just want to directly mark everything as `evict_first` for persistent reductions and pointwise ops.

Now, what about foreach kernels and split scans?

**Review by lezcano — COMMENTED** [review]
Yep, similar to what @Chillee already said. I wrote the current algorithm, and it pretty much centres around setting `evict_last` correctly when we have loops.

**Comment by shunting314** [comment]
The perf test [result](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2005%20Feb%202024%2023%3A32%3A38%20GMT&stopTime=Mon%2C%2012%20Feb%202024%2023%3A32%3A38%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/96/head&lCommit=ed78b3a653180c0f7ef903de2d3af9909517d2de&rBranch=main&rCommit=699ae72f51b2c9bc4e5b21340da68f4a732d8528) may not be accurate because of other changes get mixed in. I run one more rebase on a nightly commit.

**Comment by shunting314** [comment]
Hmm, here is the latest accurate (since I rebase on a nightly commit) [perf result](https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2006%20Feb%202024%2019%3A41%3A49%20GMT&stopTime=Tue%2C%2013%20Feb%202024%2019%3A41%3A49%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/96/head&lCommit=e8eae859f4ae21a33b29e86b112e00fbbf2f8b2e&rBranch=main&rCommit=8a3c2410946a0474909aacea02a9372f82aa109c)

it's interesting that HF has 3% regress and other benchmark suites are neutral..

I tried to repro the regress of ElectraForQuestionAnswering (so that I can check which kernel get slowed down) locally. This model regress from 3.35x to 3.25x. But on my devserver, it's neutral 3.212x to 3.214x. So I can not really figure out which kernel get slowed down...

I think overall maybe the 'evict_first' heuristic only helps for very specific kernels but doesn't help for most others.

I'll suspend this PR for now unless people have other suggestions.

**Comment by Chillee** [comment]
@shunting314 I'm not actually sure it regressed - it looks like "debertaforquestionanswering" is now failing? I would perhaps do another run haha.

**Comment by shunting314** [comment]
> it looks like "debertaforquestionanswering" is now failing? I would perhaps do another run haha.

The failure shows up in my earlier tests as well. Should be real. yea, I think I can at least fix it first and retry the perf test.

**Comment by shunting314** [comment]
> I'm not actually sure it regressed - it looks like "debertaforquestionanswering" is now failing? I would perhaps do another run haha.
```
>>> (1.97 ** 44 * 3.01) ** (1.0/45)
1.9886453692945196
```
I did some simple math here. Even with debertaforquestionanswering  accuracy fixed (assume 3.01x speedup as shown on the dashboard), we are go to have around 1.99x speedup for the all 45 models (44 plus debertaforquestionanswering ). Still no gains overall :)

But the accuracy failure for debertaforquestionanswering may reveal some issues that worth investigating
